src/profile/components/basicData/BasicData.tsx

```diff
@@ -1,28 +1,172 @@
-import React from 'react';
+import React, { useState } from 'react';
 import { useTranslation } from 'react-i18next';
+import { Field, Formik, FormikProps, Form } from 'formik';
+import { TextInput } from 'hds-react';
+import to from 'await-to-js';
 
 import commonFormStyles from '../../../common/cssHelpers/form.module.css';
 import LabeledValue from '../../../common/labeledValue/LabeledValue';
 import ProfileSection from '../../../common/profileSection/ProfileSection';
 import { getFormFields } from '../../helpers/formProperties';
-import { basicDataType, BasicDataValue } from '../../helpers/editData';
-import { useProfileDataEditor } from '../../hooks/useProfileDataEditor';
+import {
+  basicDataType,
+  BasicDataValue,
+  EditDataValue,
+} from '../../helpers/editData';
+import {
+  ActionListener,
+  saveTypeToAction,
+  useProfileDataEditor,
+} from '../../hooks/useProfileDataEditor';
+import { basicDataSchema } from '../../../common/schemas/schemas';
+import { getFieldError, getIsInvalid } from '../../helpers/formik';
+import useNotificationContent from '../editingNotifications/useNotificationContent';
+import EditingNotifications from '../editingNotifications/EditingNotifications';
+import EditButtons, { ActionHandler } from '../editButtons/EditButtons';
+import FormButtons from '../formButtons/FormButtons';
+import SaveIndicator from '../saveIndicator/SaveIndicator';
+
+type FormikValues = BasicDataValue;
 
 function BasicData(): React.ReactElement | null {
-  const { editDataList } = useProfileDataEditor({
+  const { editDataList, save, reset } = useProfileDataEditor({
     dataType: basicDataType,
   });
+  const [isEditing, setEditing] = useState(false);
   const { t } = useTranslation();
+  const {
+    content,
+    setErrorMessage,
+    setSuccessMessage,
+    clearMessage,
+  } = useNotificationContent();
   const testId = basicDataType;
 
   if (!editDataList || !editDataList[0]) {
     return null;
   }
   const editData = editDataList[0];
-  const { value } = editData;
+  const { value, saving } = editData;
   const { firstName, nickname, lastName } = value as BasicDataValue;
   const formFields = getFormFields(basicDataType);
 
+  const hasFieldError = (
+    formikProps: FormikProps<FormikValues>,
+    fieldName: keyof FormikValues
+  ): boolean => getIsInvalid<FormikValues>(formikProps, fieldName, true);
+
+  const getFieldErrorMessage = (
```

Comment:
I find the naming of this function little problematic. It is that it does not seem to return the FieldErrorMessage if there is no fieldError? One option would be to either move the check for _hasFieldError_ outside of this function, and then only call this function when we are sure it will return the FieldErrorMessage. Other option would be to name it differently: _getFieldErrorMessageIfExists_ or _getFieldErrorMessageOrUndefined_. Those names would give hint to user of the function that there is a chance it won't return the FieldErrorMessage

Reply:
Isn't typescript return types enough to tell what function returns? :) In this case it is not written here but seen when you hover over the function. And where ever the function is called, you can see return types by hovering over the function call. This is editor specific of course, but arent' they visible in all normal editors?

I find it hard to put all return types to the function name itself.

Reply:
Ide and hovering definitely helps. But I find it easier to read if I can determine what the function does (returns in this case) from the name. Hovering over is an extra step that slows down reading, and should be only needed if I wish to know more details. I agree that a function with many return types in its name feels bad. I think in general it would be good for a function to only return one types of things. The more types the function returns, the more complex it is, and the more harder it is for the reader of the code. I would probably prefer moving the check for hasFieldError outside of this function for that reason.  But yeah obviously this is very opinion dependant when it comes to readability and what is enough.

Reply:
Two reasons why hasFieldError is in getFieldErrorMessage

1: it is used in an element-prop:
helperText={getFieldErrorMessage(formikProps, 'nickname')}

in this case it should be changed to:
helperText={hasFieldError(formikProps, 'nickname') && getFieldErrorMessage(formikProps, 'nickname')}

that's looks terrible IMHO. And that in 3 places in this file and later on in address, postal code, city, email and phone. 

2: it has always been there. It this file it looks like it was created here, but it's a copy from old code

Reply:
Should I copy the return type of getFieldError to getFieldErrorMessage? Copying code for faster readability is not very convenient. When it getFieldError changes, it must be changed everywhere. Later on also in addresses, emails and phones will use it.

I can make a common type for them, but seems a bit excessive. You cannot read it in IDE either.

Reply:
To my eye this:

```helperText={hasFieldError(formikProps, 'nickname') && getFieldErrorMessage(formikProps, 'nickname')}```

doesn't look that bad because you then know the condition when there will be a helperText. 😅 But I guess this is about what reader of the code is accustomed to. I've probably just used to read code where logic is shown in higher abstraction level. I think the current solution is good enough, we can go with that. 👍

Reply:
I left this as it is.